src/providers/swift.rs

```diff
@@ -0,0 +1,174 @@
+use super::Provider;
+use crate::nixpacks::{
+    app::App,
+    environment::Environment,
+    nix::pkg::Pkg,
+    phase::{BuildPhase, InstallPhase, SetupPhase, StartPhase},
+};
+use anyhow::{bail, Result};
+
+static DEFAULT_SWIFT_VERSION: &str = "5.4.2";
+
+// From: https://lazamar.co.uk/nix-versions/?channel=nixpkgs-unstable&package=swift
+static AVAILABLE_SWIFT_VERSIONS: &[(&str, &str)] = &[
+    ("3.1", "aeaa79dc82980869a88a5955ea3cd3e1944b7d80"),
+    ("3.1.1", "8414d8386b9a6b855b291fb3f01a4e3b04c08bbb"),
+    ("4.0.3", "2c9d2d65266c2c3aca1e4c80215de8bee5295b04"),
+    ("4.1", "92a047a6c4d46a222e9c323ea85882d0a7a13af8"),
+    ("4.1.3", "a3962299f14944a0e9ccf8fd84bd7be524b74cd6"),
+    ("4.2.1", "7ff8a16f0726342f0a25697867d8c1306d4da7b0"),
+    ("4.2.3", "3fa154fd7fed3d6a94322bf08a6def47d6f8e0f6"),
+    ("5.0.1", "4599f2bb9a5a6b1482e72521ead95cb24e0aa819"),
+    ("5.0.2", "a9eb3eed170fa916e0a8364e5227ee661af76fde"),
+    ("5.1.1", "9986226d5182c368b7be1db1ab2f7488508b5a87"),
+    ("5.4", "c82b46413401efa740a0b994f52e9903a4f6dcd5"),
+    ("5.4.2", "c82b46413401efa740a0b994f52e9903a4f6dcd5"),
+];
+
+pub struct SwiftProvider {}
+
+impl Provider for SwiftProvider {
+    fn name(&self) -> &str {
+        "swift"
+    }
+
+    fn detect(&self, app: &App, _env: &Environment) -> Result<bool> {
+        Ok(app.includes_file("Package.swift"))
+    }
+
+    fn setup(&self, app: &App, _env: &Environment) -> Result<Option<SetupPhase>> {
+        let mut setup_phase = SetupPhase::new(vec![
+            Pkg::new("coreutils"),
+            Pkg::new("swift"),
+            Pkg::new("clang"),
+            Pkg::new("zlib"),
+            Pkg::new("zlib.dev"),
+        ]);
+
+        let swift_version = SwiftProvider::get_swift_version(app)?;
+        let rev = SwiftProvider::version_number_to_rev(&swift_version);
+
+        if let Some(rev) = rev {
+            setup_phase.set_archive(rev);
+        } else {
+            setup_phase.set_archive("c82b46413401efa740a0b994f52e9903a4f6dcd5".to_string());
```

Comment:
Can this be pulled from the `AVAILABLE_SWIFT_VERSIONS` above so it is not hardcoded

Reply:
done